File: contracts/ts/cache.ts
/**
 * ICacheService provides a key-value cache for sessions, rate limiting, and hot data.
 * Required by the auth module (session storage) and jobs module (queue backend).
 */
export interface ICacheService {
  /**
   * Retrieves the raw bytes for a key.
   * Returns undefined if the key does not exist or has expired.
   */
  get(key: string): Promise<Buffer | undefined>;

  /**
   * Stores a value with an optional TTL in seconds.
   * If ttlSeconds is 0 or omitted, the key does not expire.
   */
  set(key: string, value: Buffer | string, ttlSeconds?: number): Promise<void>;

  /**
   * Removes a key. Returns without error if the key did not exist.
   */
  delete(key: string): Promise<void>;

  /**
   * Checks whether a key exists without retrieving its value.
   */
  exists(key: string): Promise<boolean>;

  /**
   * Atomically increments an integer value by delta (default: 1).
   * Creates the key with value delta if it does not exist.
   * Returns the new value after incrementing.
   */
  increment(key: string, delta?: number): Promise<number>;

  /**
   * Sets a key only if it does not already exist (atomic SET NX).
   * Returns true if the key was set, false if it already existed.
   * Use for distributed locks and deduplication.
   */
  setNX(key: string, value: Buffer | string, ttlSeconds?: number): Promise<boolean>;

  /**
   * Updates the TTL for an existing key.
   * Throws CacheError with code NOT_FOUND if the key does not exist.
   */
  expire(key: string, ttlSeconds: number): Promise<void>;

  /**
   * Deletes all keys matching the given glob pattern.
   * Use with care — can be slow on large keyspaces.
   */
  flushPattern(pattern: string): Promise<void>;
}

File: contracts/go/cache.go
package contracts

import (
	"context"
	"time"
)

// CacheService provides a key-value cache for sessions, rate limiting, and hot data.
// Required by the auth module (session storage) and jobs module (queue backend).
type CacheService interface {
	// Get retrieves the value for a key.
	// Returns ErrNotFound if the key does not exist or has expired.
	Get(ctx context.Context, key string) ([]byte, error)

	// Set stores a value with an optional TTL.
	// If ttl is 0, the key does not expire.
	Set(ctx context.Context, key string, value []byte, ttl time.Duration) error

	// Delete removes a key. Returns nil if the key did not exist.
	Delete(ctx context.Context, key string) error

	// Exists checks whether a key exists (without retrieving its value).
	Exists(ctx context.Context, key string) (bool, error)

	// Increment atomically increments an integer value by delta.
	// Creates the key with value delta if it does not exist.
	// Returns the new value after incrementing.
	Increment(ctx context.Context, key string, delta int64) (int64, error)

	// SetNX sets a key only if it does not already exist (atomic).
	// Returns true if the key was set, false if it already existed.
	// Use for distributed locks and deduplication.
	SetNX(ctx context.Context, key string, value []byte, ttl time.Duration) (bool, error)

	// Expire updates the TTL for an existing key.
	// Returns ErrNotFound if the key does not exist.
	Expire(ctx context.Context, key string, ttl time.Duration) error

	// FlushPattern deletes all keys matching the given glob pattern.
	// Use with care — can be slow on large keyspaces.
	FlushPattern(ctx context.Context, pattern string) error
}
